planner/core/task.go

```diff
@@ -557,28 +557,186 @@ func (p *PhysicalHashJoin) attach2Task(tasks ...task) task {
 	return task
 }
 
+// for different decimal scale and precision, tiflash uses different underlying type.
+// Here we check the scale and precision to decide whether conversion is a must.
+func needDecimalConvert(tp *types.FieldType, dec int, l int) bool {
+	if tp.Decimal != dec {
+		return true
+	}
+	if tp.Flen >= 0 && tp.Flen <= 9 && l >= 0 && l <= 9 {
+		return false
+	}
+	if tp.Flen > 9 && tp.Flen <= 18 && l > 9 && l <= 18 {
+		return false
+	}
+	if tp.Flen > 18 && tp.Flen <= 38 && l > 18 && l <= 38 {
+		return false
+	}
+	if tp.Flen > 38 && tp.Flen <= 65 && l > 38 && l <= 65 {
+		return false
+	}
+	return true
+}
+
+func negotiateCommonType(lType, rType *types.FieldType) (*types.FieldType, bool, bool) {
+	lExtend := 0
+	rExtend := 0
+	cDec := rType.Decimal
+	if lType.Decimal < rType.Decimal {
+		lExtend = rType.Decimal - lType.Decimal
+	} else if lType.Decimal > rType.Decimal {
+		rExtend = lType.Decimal - rType.Decimal
+		cDec = lType.Decimal
+	}
+	lLen, rLen := lType.Flen+lExtend, rType.Flen+rExtend
+	cLen := lLen
+	if rLen > cLen {
+		cLen = rLen
+	}
+	if cLen > 65 {
+		cLen = 65
+	}
```

Comment:
cLen := max(65, lType.Flen+lExtend, rType.Flen+rExtend)?

Reply:
Is there a max function in golang?

Reply:
myMax https://github.com/pingcap/tidb/blob/8a3705ee404031ad9ad04a5b9587cb9ce6a6446e/types/helper.go#L109

Reply:
It's not exported so that I can't invoke it.

Reply:
https://github.com/pingcap/tidb/blob/31c13d6e3893ad8dd120cc850fd4ed0f632edb20/expression/builtin_arithmetic.go#L887